gooddata-java/src/main/java/com/gooddata/sdk/service/GoodDataServices.java

```diff
@@ -0,0 +1,189 @@
+    private final MetadataService metadataService;
+    private final ModelService modelService;
+    private final GdcService gdcService;
+    private final DataStoreService dataStoreService;
+    private final DatasetService datasetService;
+    private final ConnectorService connectorService;
+    private final ProcessService processService;
+    private final WarehouseService warehouseService;
+    private final NotificationService notificationService;
+    private final ExportImportService exportImportService;
+    private final FeatureFlagService featureFlagService;
+    private final OutputStageService outputStageService;
+    private final ProjectTemplateService projectTemplateService;
+    private final ExportService exportService;
+    private final AuditEventService auditEventService;
+    private final ExecuteAfmService executeAfmService;
+    private final LcmService lcmService;
+
+    GoodDataServices(final GoodDataRestProvider goodDataRestProvider) {
+        this.goodDataRestProvider = goodDataRestProvider;
+
+        accountService = new AccountService(getRestTemplate(), getSettings());
+        projectService = new ProjectService(getRestTemplate(), accountService, getSettings());
+        metadataService = new MetadataService(getRestTemplate(), getSettings());
+        modelService = new ModelService(getRestTemplate(), getSettings());
+        gdcService = new GdcService(getRestTemplate(), getSettings());
+        exportService = new ExportService(getRestTemplate(), getSettings());
+        warehouseService = new WarehouseService(getRestTemplate(), getSettings());
+        connectorService = new ConnectorService(getRestTemplate(), projectService, getSettings());
+        notificationService = new NotificationService(getRestTemplate(), getSettings());
+        exportImportService = new ExportImportService(getRestTemplate(), getSettings());
+        featureFlagService = new FeatureFlagService(getRestTemplate(), getSettings());
+        outputStageService = new OutputStageService(getRestTemplate(), getSettings());
+        projectTemplateService = new ProjectTemplateService(getRestTemplate(), getSettings());
+        auditEventService = new AuditEventService(getRestTemplate(), accountService, getSettings());
+        executeAfmService = new ExecuteAfmService(getRestTemplate(), getSettings());
+        lcmService = new LcmService(getRestTemplate(), getSettings());
+
+        dataStoreService = initDataStoreService(goodDataRestProvider, gdcService);
+
+        datasetService = new DatasetService(getRestTemplate(), dataStoreService, getSettings());
+        processService = new ProcessService(getRestTemplate(), accountService, dataStoreService, getSettings());
+    }
+
+    private static DataStoreService initDataStoreService(final GoodDataRestProvider goodDataRestProvider, final GdcService gdcService) {
+        if (goodDataRestProvider instanceof SingleEndpointGoodDataRestProvider) {
```

Comment:
This is sort of ugly. What about using inheritance and overridden methods instead?

Reply:
Not sure what's on your mind. The thing is that `DataStoreService` is and probably always be dependent on this particular class. If I am missing something please be more specific.

BTW: I agree it's ugly, but the ugliness comes mostly from the nature of Sardine (http client) dependent `DataStoreService`.